// @sap/ux-specification/dist/types/src/common/Application.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlexibleColumnLayoutAggregations = exports.StatePreservationMode = exports.TableColumnVerticalAlignment = exports.DraftDiscardEnabledSettings = void 0;
var DraftDiscardEnabledSettings;
(function (DraftDiscardEnabledSettings) {
    DraftDiscardEnabledSettings["restricted"] = "restricted";
})(DraftDiscardEnabledSettings || (exports.DraftDiscardEnabledSettings = DraftDiscardEnabledSettings = {}));
var TableColumnVerticalAlignment;
(function (TableColumnVerticalAlignment) {
    TableColumnVerticalAlignment["Top"] = "Top";
    TableColumnVerticalAlignment["Middle"] = "Middle";
    TableColumnVerticalAlignment["Bottom"] = "Bottom";
})(TableColumnVerticalAlignment || (exports.TableColumnVerticalAlignment = TableColumnVerticalAlignment = {}));
var StatePreservationMode;
(function (StatePreservationMode) {
    StatePreservationMode["persistence"] = "persistence";
    StatePreservationMode["discovery"] = "discovery";
})(StatePreservationMode || (exports.StatePreservationMode = StatePreservationMode = {}));
var FlexibleColumnLayoutAggregations;
(function (FlexibleColumnLayoutAggregations) {
    FlexibleColumnLayoutAggregations["BeginColumnPages"] = "beginColumnPages";
    FlexibleColumnLayoutAggregations["MidColumnPages"] = "midColumnPages";
    FlexibleColumnLayoutAggregations["EndColumnPages"] = "endColumnPages";
})(FlexibleColumnLayoutAggregations || (exports.FlexibleColumnLayoutAggregations = FlexibleColumnLayoutAggregations = {}));
