// @sap/ux-specification/dist/types/src/common/color.js
"use strict";
//Copy from https://ui5.sap.com/resources/sap/viz/library-dbg.js as enum
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChartColor = void 0;
var ChartColor;
(function (ChartColor) {
    // Qualitative color
    ChartColor["sapUiChartPaletteQualitativeHue1"] = "sapUiChartPaletteQualitativeHue1";
    ChartColor["sapUiChartPaletteQualitativeHue2"] = "sapUiChartPaletteQualitativeHue2";
    ChartColor["sapUiChartPaletteQualitativeHue3"] = "sapUiChartPaletteQualitativeHue3";
    ChartColor["sapUiChartPaletteQualitativeHue4"] = "sapUiChartPaletteQualitativeHue4";
    ChartColor["sapUiChartPaletteQualitativeHue5"] = "sapUiChartPaletteQualitativeHue5";
    ChartColor["sapUiChartPaletteQualitativeHue6"] = "sapUiChartPaletteQualitativeHue6";
    ChartColor["sapUiChartPaletteQualitativeHue7"] = "sapUiChartPaletteQualitativeHue7";
    ChartColor["sapUiChartPaletteQualitativeHue8"] = "sapUiChartPaletteQualitativeHue8";
    ChartColor["sapUiChartPaletteQualitativeHue9"] = "sapUiChartPaletteQualitativeHue9";
    ChartColor["sapUiChartPaletteQualitativeHue10"] = "sapUiChartPaletteQualitativeHue10";
    ChartColor["sapUiChartPaletteQualitativeHue11"] = "sapUiChartPaletteQualitativeHue11";
    ChartColor["sapUiChartPaletteQualitativeHue12"] = "sapUiChartPaletteQualitativeHue12";
    ChartColor["sapUiChartPaletteQualitativeHue13"] = "sapUiChartPaletteQualitativeHue13";
    ChartColor["sapUiChartPaletteQualitativeHue14"] = "sapUiChartPaletteQualitativeHue14";
    ChartColor["sapUiChartPaletteQualitativeHue15"] = "sapUiChartPaletteQualitativeHue15";
    ChartColor["sapUiChartPaletteQualitativeHue16"] = "sapUiChartPaletteQualitativeHue16";
    ChartColor["sapUiChartPaletteQualitativeHue17"] = "sapUiChartPaletteQualitativeHue17";
    ChartColor["sapUiChartPaletteQualitativeHue18"] = "sapUiChartPaletteQualitativeHue18";
    ChartColor["sapUiChartPaletteQualitativeHue19"] = "sapUiChartPaletteQualitativeHue19";
    ChartColor["sapUiChartPaletteQualitativeHue20"] = "sapUiChartPaletteQualitativeHue20";
    ChartColor["sapUiChartPaletteQualitativeHue21"] = "sapUiChartPaletteQualitativeHue21";
    ChartColor["sapUiChartPaletteQualitativeHue22"] = "sapUiChartPaletteQualitativeHue22";
    // Semantic color
    ChartColor["sapUiChartPaletteSemanticBadLight3"] = "sapUiChartPaletteSemanticBadLight3";
    ChartColor["sapUiChartPaletteSemanticBadLight2"] = "sapUiChartPaletteSemanticBadLight2";
    ChartColor["sapUiChartPaletteSemanticBadLight1"] = "sapUiChartPaletteSemanticBadLight1";
    ChartColor["sapUiChartPaletteSemanticBad"] = "sapUiChartPaletteSemanticBad";
    ChartColor["sapUiChartPaletteSemanticBadDark1"] = "sapUiChartPaletteSemanticBadDark1";
    ChartColor["sapUiChartPaletteSemanticBadDark2"] = "sapUiChartPaletteSemanticBadDark2";
    ChartColor["sapUiChartPaletteSemanticCriticalLight3"] = "sapUiChartPaletteSemanticCriticalLight3";
    ChartColor["sapUiChartPaletteSemanticCriticalLight2"] = "sapUiChartPaletteSemanticCriticalLight2";
    ChartColor["sapUiChartPaletteSemanticCriticalLight1"] = "sapUiChartPaletteSemanticCriticalLight1";
    ChartColor["sapUiChartPaletteSemanticCritical"] = "sapUiChartPaletteSemanticCritical";
    ChartColor["sapUiChartPaletteSemanticCriticalDark1"] = "sapUiChartPaletteSemanticCriticalDark1";
    ChartColor["sapUiChartPaletteSemanticCriticalDark2"] = "sapUiChartPaletteSemanticCriticalDark2";
    ChartColor["sapUiChartPaletteSemanticGoodLight3"] = "sapUiChartPaletteSemanticGoodLight3";
    ChartColor["sapUiChartPaletteSemanticGoodLight2"] = "sapUiChartPaletteSemanticGoodLight2";
    ChartColor["sapUiChartPaletteSemanticGoodLight1"] = "sapUiChartPaletteSemanticGoodLight1";
    ChartColor["sapUiChartPaletteSemanticGood"] = "sapUiChartPaletteSemanticGood";
    ChartColor["sapUiChartPaletteSemanticGoodDark1"] = "sapUiChartPaletteSemanticGoodDark1";
    ChartColor["sapUiChartPaletteSemanticGoodDark2"] = "sapUiChartPaletteSemanticGoodDark2";
    ChartColor["sapUiChartPaletteSemanticNeutralLight3"] = "sapUiChartPaletteSemanticNeutralLight3";
    ChartColor["sapUiChartPaletteSemanticNeutralLight2"] = "sapUiChartPaletteSemanticNeutralLight2";
    ChartColor["sapUiChartPaletteSemanticNeutralLight1"] = "sapUiChartPaletteSemanticNeutralLight1";
    ChartColor["sapUiChartPaletteSemanticNeutral"] = "sapUiChartPaletteSemanticNeutral";
    ChartColor["sapUiChartPaletteSemanticNeutralDark1"] = "sapUiChartPaletteSemanticNeutralDark1";
    ChartColor["sapUiChartPaletteSemanticNeutralDark2"] = "sapUiChartPaletteSemanticNeutralDark2";
    ChartColor["sapUiChartPaletteNoSemDiv1Dark2"] = "sapUiChartPaletteNoSemDiv1Dark2";
    ChartColor["sapUiChartPaletteNoSemDiv1Dark1"] = "sapUiChartPaletteNoSemDiv1Dark1";
    ChartColor["sapUiChartPaletteNoSemDiv1"] = "sapUiChartPaletteNoSemDiv1";
    ChartColor["sapUiChartPaletteNoSemDiv1Light1"] = "sapUiChartPaletteNoSemDiv1Light1";
    ChartColor["sapUiChartPaletteNoSemDiv1Light2"] = "sapUiChartPaletteNoSemDiv1Light2";
    ChartColor["sapUiChartPaletteNoSemDiv1Light3"] = "sapUiChartPaletteNoSemDiv1Light3";
    // Sequential color
    ChartColor["sapUiChartPaletteSequentialHue1Light3"] = "sapUiChartPaletteSequentialHue1Light3";
    ChartColor["sapUiChartPaletteSequentialHue1Light2"] = "sapUiChartPaletteSequentialHue1Light2";
    ChartColor["sapUiChartPaletteSequentialHue1Light1"] = "sapUiChartPaletteSequentialHue1Light1";
    ChartColor["sapUiChartPaletteSequentialHue1"] = "sapUiChartPaletteSequentialHue1";
    ChartColor["sapUiChartPaletteSequentialHue1Dark1"] = "sapUiChartPaletteSequentialHue1Dark1";
    ChartColor["sapUiChartPaletteSequentialHue1Dark2"] = "sapUiChartPaletteSequentialHue1Dark2";
    ChartColor["sapUiChartPaletteSequentialHue2Light3"] = "sapUiChartPaletteSequentialHue2Light3";
    ChartColor["sapUiChartPaletteSequentialHue2Light2"] = "sapUiChartPaletteSequentialHue2Light2";
    ChartColor["sapUiChartPaletteSequentialHue2Light1"] = "sapUiChartPaletteSequentialHue2Light1";
    ChartColor["sapUiChartPaletteSequentialHue2"] = "sapUiChartPaletteSequentialHue2";
    ChartColor["sapUiChartPaletteSequentialHue2Dark1"] = "sapUiChartPaletteSequentialHue2Dark1";
    ChartColor["sapUiChartPaletteSequentialHue2Dark2"] = "sapUiChartPaletteSequentialHue2Dark2";
    ChartColor["sapUiChartPaletteSequentialHue3Light3"] = "sapUiChartPaletteSequentialHue3Light3";
    ChartColor["sapUiChartPaletteSequentialHue3Light2"] = "sapUiChartPaletteSequentialHue3Light2";
    ChartColor["sapUiChartPaletteSequentialHue3Light1"] = "sapUiChartPaletteSequentialHue3Light1";
    ChartColor["sapUiChartPaletteSequentialHue3"] = "sapUiChartPaletteSequentialHue3";
    ChartColor["sapUiChartPaletteSequentialHue3Dark1"] = "sapUiChartPaletteSequentialHue3Dark1";
    ChartColor["sapUiChartPaletteSequentialHue3Dark2"] = "sapUiChartPaletteSequentialHue3Dark2";
    ChartColor["sapUiChartPaletteSequentialHue6Light3"] = "sapUiChartPaletteSequentialHue6Light3";
    ChartColor["sapUiChartPaletteSequentialHue6Light2"] = "sapUiChartPaletteSequentialHue6Light2";
    ChartColor["sapUiChartPaletteSequentialHue6Light1"] = "sapUiChartPaletteSequentialHue6Light1";
    ChartColor["sapUiChartPaletteSequentialHue6"] = "sapUiChartPaletteSequentialHue6";
    ChartColor["sapUiChartPaletteSequentialHue6Dark1"] = "sapUiChartPaletteSequentialHue6Dark1";
    ChartColor["sapUiChartPaletteSequentialHue6Dark2"] = "sapUiChartPaletteSequentialHue6Dark2";
    ChartColor["sapUiChartPaletteSequentialNeutralLight3"] = "sapUiChartPaletteSequentialNeutralLight3";
    ChartColor["sapUiChartPaletteSequentialNeutralLight2"] = "sapUiChartPaletteSequentialNeutralLight2";
    ChartColor["sapUiChartPaletteSequentialNeutralLight1"] = "sapUiChartPaletteSequentialNeutralLight1";
    ChartColor["sapUiChartPaletteSequentialNeutral"] = "sapUiChartPaletteSequentialNeutral";
    ChartColor["sapUiChartPaletteSequentialNeutralDark1"] = "sapUiChartPaletteSequentialNeutralDark1";
    ChartColor["sapUiChartPaletteSequentialNeutralDark2"] = "sapUiChartPaletteSequentialNeutralDark2";
    ChartColor["sapUiChoroplethRegionBG"] = "sapUiChoroplethRegionBG";
    ChartColor["sapUiChartZeroAxisColor"] = "sapUiChartZeroAxisColor";
    ChartColor["sapUiNegativeElement"] = "sapUiNegativeElement";
    ChartColor["sapUiCriticalElement"] = "sapUiCriticalElement";
    ChartColor["sapUiPositiveElement"] = "sapUiPositiveElement";
    ChartColor["sapUiNeutralElement"] = "sapUiNeutralElement";
})(ChartColor || (exports.ChartColor = ChartColor = {}));
