// @sap/ux-specification/dist/specification/src/sync/v4/export/controls/ObjectPageCustomSection.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPageCustomSubSectionFragment = exports.ObjectPageCustomSectionFragment = exports.ObjectPageCustomSectionActions = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
class ObjectPageCustomSectionActions {
}
exports.ObjectPageCustomSectionActions = ObjectPageCustomSectionActions;
class ObjectPageCustomSectionFragment {
}
exports.ObjectPageCustomSectionFragment = ObjectPageCustomSectionFragment;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId) => {
                return (0, application_1.getCustomSectionPath)(pageName, manifest, sectionId, 'position');
            },
            key: 'anchor'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "relatedFacet", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId) => (0, application_1.getCustomSectionPath)(pageName, manifest, sectionId, 'position'),
            key: 'placement'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "relativePosition", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "title", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: (manifestSection) => {
                // During import, we need to consider both `name` and `template` to correctly fill our config and to be upgrade safe
                return manifestSection.name || manifestSection.template;
            },
            key: 'template' // key: 'name' deprecated since '1.86'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "fragmentName", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: false,
            export: false
        },
        xml: (config) => {
            return config['fragmentName'];
        }
    }),
    (0, decorators_1.validity)({
        since: '1.96.0'
    })
], ObjectPageCustomSectionFragment.prototype, "controls", void 0);
class ObjectPageCustomSubSectionFragment extends ObjectPageCustomSectionFragment {
}
exports.ObjectPageCustomSubSectionFragment = ObjectPageCustomSubSectionFragment;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (...args) => {
                const resolvedPath = application_1.getCustomSubSectionPath.apply({}, args);
                return `${resolvedPath}/position`;
            },
            key: 'anchor'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "relatedFacet", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (...args) => {
                const resolvedPath = application_1.getCustomSubSectionPath.apply({}, args);
                return `${resolvedPath}/position`;
            },
            key: 'placement'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "relativePosition", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSubSectionPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "title", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSubSectionPath,
            import: (manifestSection) => {
                // During import, we need to consider both `name` and `template` to correctly fill our config and to be upgrade safe
                return manifestSection.name || manifestSection.template;
            },
            key: 'template' // key: 'name' deprecated since '1.86'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "fragmentName", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSubSectionPath,
            import: false,
            export: false
        },
        xml: (config) => {
            return config['fragmentName'];
        }
    }),
    (0, decorators_1.validity)({
        since: '1.96.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "controls", void 0);
