// @sap/ux-specification/dist/specification/src/sync/v4/export/controls/ObjectPageSectionAction.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomSectionActionOP = exports.CustomSectionActionPositionOP = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const control_1 = require("../../../common/decoration/control");
const application_1 = require("../../application");
/**
 * Sync class for Position
 */
class CustomSectionActionPositionOP {
}
exports.CustomSectionActionPositionOP = CustomSectionActionPositionOP;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, actionId) => (0, application_1.getCustomSectionActionPath)(pageName, manifest, sectionId, actionId) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomSectionActionPositionOP.prototype, "anchor", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, actionId) => (0, application_1.getCustomSectionActionPath)(pageName, manifest, sectionId, actionId) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomSectionActionPositionOP.prototype, "placement", void 0);
class CustomSectionActionOP {
}
exports.CustomSectionActionOP = CustomSectionActionOP;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionActionPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomSectionActionOP.prototype, "text", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionActionPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomSectionActionOP.prototype, "position", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionActionPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomSectionActionOP.prototype, "press", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionActionPath
        },
        generate: control_1.addPatternForBindingChangeOfEnumOPV4
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomSectionActionOP.prototype, "visible", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionActionPath
        },
        generate: control_1.addPatternForBindingChangeOfEnumOPV4
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomSectionActionOP.prototype, "enabled", void 0);
