// @sap/ux-cds-compiler-facade/dist/annotationAssignmentToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnotationAssignmentTokenImpl = exports.NEWLINE_REGEX = void 0;
const odata_annotation_core_types_1 = require("@sap-ux/odata-annotation-core-types");
const genericToken_1 = require("./genericToken");
exports.NEWLINE_REGEX = /\r?\n/g;
class AnnotationAssignmentTokenImpl extends genericToken_1.GenericTokenImpl {
    constructor(uri, tokenIndex, text, line, character, carrierName, carrier, carrierNameRange, annotationValue) {
        super(uri, tokenIndex, text, line, character, annotationValue);
        this.uri = uri;
        this.tokenIndex = tokenIndex;
        this.text = text;
        this.line = line;
        this.character = character;
        this.carrierName = carrierName;
        this.carrier = carrier;
        this.carrierNameRange = carrierNameRange;
        this.annotationValue = annotationValue;
    }
    get range() {
        if (!this._range) {
            const lines = this.text.split(exports.NEWLINE_REGEX);
            this._range = odata_annotation_core_types_1.Range.create(this.line, this.character, this.line + lines.length - 1, lines.length === 1 ? this.character + this.text.length : lines.slice(-1)[0].length);
        }
        return this._range;
    }
}
exports.AnnotationAssignmentTokenImpl = AnnotationAssignmentTokenImpl;
