// @sap/ux-cds-compiler-facade/dist/cds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFileUriMap = exports.severityMap = void 0;
const url_1 = require("url");
const odata_annotation_core_types_1 = require("@sap-ux/odata-annotation-core-types");
exports.severityMap = {
    Debug: odata_annotation_core_types_1.DiagnosticSeverity.Hint,
    Error: odata_annotation_core_types_1.DiagnosticSeverity.Error,
    Warning: odata_annotation_core_types_1.DiagnosticSeverity.Warning,
    Info: odata_annotation_core_types_1.DiagnosticSeverity.Information
};
function getFileUriMap(ast) {
    const fileUriMap = new Map();
    if (!ast?.sources) {
        return fileUriMap;
    }
    Object.keys(ast.sources).forEach((sourceUri) => {
        const relativeUri = ast.sources[sourceUri].location?.file;
        const absoluteUri = (0, url_1.pathToFileURL)(sourceUri).toString();
        if (relativeUri) {
            fileUriMap.set(relativeUri, absoluteUri.toString());
        }
    });
    return fileUriMap;
}
exports.getFileUriMap = getFileUriMap;
