// @sap/ux-cds-compiler-facade/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCdsCompilerFacadeForRoot = exports.getCdsFiles = exports.createCdsCompilerFacade = exports.getCdsArtifacts = exports.createMetadataCollector = exports.getMetadataElementsFromMap = void 0;
var metadata_1 = require("./metadata");
Object.defineProperty(exports, "getMetadataElementsFromMap", { enumerable: true, get: function () { return metadata_1.getMetadataElementsFromMap; } });
Object.defineProperty(exports, "createMetadataCollector", { enumerable: true, get: function () { return metadata_1.createMetadataCollector; } });
var cdsArtifacts_1 = require("./cdsArtifacts");
Object.defineProperty(exports, "getCdsArtifacts", { enumerable: true, get: function () { return cdsArtifacts_1.getCdsArtifacts; } });
var facade_1 = require("./facade");
Object.defineProperty(exports, "createCdsCompilerFacade", { enumerable: true, get: function () { return facade_1.createCdsCompilerFacade; } });
var compile_1 = require("./compile");
Object.defineProperty(exports, "getCdsFiles", { enumerable: true, get: function () { return compile_1.getCdsFiles; } });
Object.defineProperty(exports, "createCdsCompilerFacadeForRoot", { enumerable: true, get: function () { return compile_1.createCdsCompilerFacadeForRoot; } });
