// @sap-ux/fiori-annotation-api/dist/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compareByRange = exports.annotationReferenceToString = exports.increaseIndent = exports.getGenericNodeFromPointer = exports.PRIMITIVE_TYPE_NAMES = exports.pathFromUri = void 0;
var path_1 = require("./path");
Object.defineProperty(exports, "pathFromUri", { enumerable: true, get: function () { return path_1.pathFromUri; } });
var constants_1 = require("./constants");
Object.defineProperty(exports, "PRIMITIVE_TYPE_NAMES", { enumerable: true, get: function () { return constants_1.PRIMITIVE_TYPE_NAMES; } });
var pointer_1 = require("./pointer");
Object.defineProperty(exports, "getGenericNodeFromPointer", { enumerable: true, get: function () { return pointer_1.getGenericNodeFromPointer; } });
var indent_1 = require("./indent");
Object.defineProperty(exports, "increaseIndent", { enumerable: true, get: function () { return indent_1.increaseIndent; } });
var reference_1 = require("./reference");
Object.defineProperty(exports, "annotationReferenceToString", { enumerable: true, get: function () { return reference_1.annotationReferenceToString; } });
var range_1 = require("./range");
Object.defineProperty(exports, "compareByRange", { enumerable: true, get: function () { return range_1.compareByRange; } });
