// @sap/ux-specification/dist/specification/scripts/documentation/components/ui/search.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupInBuiltSearch = exports.expandNodePath = void 0;
const toggle_1 = require("./toggle");
const utils_1 = require("./utils");
// function to create the path to nodes for search
function getNodePath(node, path) {
    const p = path || [], parent = node.parent;
    if (parent) {
        const children = parent.children || parent.data._children;
        p.unshift(children.indexOf(node));
        return getNodePath(parent, p);
    }
    else
        return p;
}
// function to set up the search
function createSearchList(root) {
    const searchList = [];
    (0, utils_1.visit)(root, 
    // creates the search list with paths to the nodes
    (current) => {
        if (current)
            searchList.push(current.data.name + '|' + getNodePath(current, []).join('-'));
    }, (current) => {
        return current.children || current.data._children;
    });
    // removes name of floorplan from search list
    searchList.splice(searchList.indexOf(root.data.name + '|'), 1).sort();
    return searchList;
}
function getDOMElement(selector) {
    return document.querySelector(selector);
}
function showDOMElement(element) {
    element.style.display = 'block';
}
function hideDOMElement(element) {
    element.style.display = 'none';
}
function removeDOMElements(container) {
    while (container.firstChild) {
        container.removeChild(container.firstChild);
    }
    hideDOMElement(container);
}
function createDOMElements(content, container) {
    const listElement = document.createElement('li');
    listElement.setAttribute('data-path', content.split('|')[1]);
    listElement.textContent = (0, utils_1.splitName)(content.split('|')[0]);
    container.appendChild(listElement);
    showDOMElement(container);
}
function searchInputEventHandler(items, inputElement, icon, container) {
    const searchValue = inputElement.value.toUpperCase().trim();
    if (searchValue === null || searchValue === '') {
        removeDOMElements(container);
        hideDOMElement(icon);
    }
    else {
        for (const item of items) {
            const listValue = item.split('|')[0].toUpperCase();
            if (listValue.indexOf(searchValue) >= 0 || listValue.replace(/\s/g, '').indexOf(searchValue) >= 0) {
                createDOMElements(item, container);
            }
            showDOMElement(icon);
        }
    }
}
function expandNodePath(path, rootNode) {
    let node = rootNode; //start with root
    for (const element of path) {
        if (node.data._children)
            (0, toggle_1.expand)(node);
        node = node.children[element];
    }
    return node;
}
exports.expandNodePath = expandNodePath;
function setupInBuiltSearch(treeData) {
    const searchInputElement = getDOMElement('#search-input');
    const searchResultContainer = getDOMElement('#search-result');
    const searchContainer = getDOMElement('#search-container');
    const searchIcon = getDOMElement('#search-icon');
    const searchCloseIcon = getDOMElement('#search-close-icon');
    if (!searchContainer) {
        return;
    }
    else {
        searchCloseIcon.addEventListener('click', () => {
            searchInputElement.value = null;
            removeDOMElements(searchResultContainer);
            hideDOMElement(searchCloseIcon);
        });
        searchContainer.addEventListener('focusin', () => {
            hideDOMElement(searchIcon);
        });
        searchContainer.addEventListener('focusout', () => {
            showDOMElement(searchIcon);
        });
        searchInputElement.addEventListener('input', () => {
            removeDOMElements(searchResultContainer);
            searchInputEventHandler(createSearchList(treeData), searchInputElement, searchCloseIcon, searchResultContainer);
        });
    }
}
exports.setupInBuiltSearch = setupInBuiltSearch;
