// @sap-ux/odata-annotation-core-types/dist/diagnostics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.INVALID_PRIMITIVE_TYPE = exports.ODATA_PATH_SEPARATOR_RULE = exports.COMMON_CASE_ISSUE = exports.INCOMPLETE_PATH_WITH_COMPATIBLE_TYPES = exports.INCOMPLETE_PATH_WITH_TYPE = exports.NO_VALIDATION_FOR_SUBNODES = exports.INVALID_TYPE_TYPE = exports.INVALID_ENUM_MEMBER_TYPE = exports.INVALID_PATH_EXPRESSION = exports.IGNORE_DUPLICATE = exports.ODATA_FUNCTION_WRONG_RETURN_TYPE = exports.RECORD_COLLECTION_PATH_NOT_ALLOWED = exports.NOT_IN_APPLICABLE_TERMS_CONSTRAINT = exports.TERM_NOT_APPLICABLE = exports.MISSING_REQUIRED_VALUE_FOR_ATTRIBUTE = exports.MISSING_REQUIRED_ATTRIBUTE = exports.ATTRIBUTE_NOT_ALLOWED_HERE = exports.UN_SUPPORTED_VOCABULARY = exports.UNKNOWN_TERM = exports.IGNORE_TARGET_VALIDATION = exports.INCOMPLETE_EXPRESSION_FORWARD_SLASH = exports.INCOMPLETE_EXPRESSION_CC_FORWARD_SLASH = exports.VALUE_REQUIRED = exports.MISSING_I18N_KEY = exports.NAME_CASE_ISSUE_PATH_VALUE = exports.NO_UNDEFINED_NAMESPACE_TYPE = exports.NO_UNUSED_NAMESPACE_TYPE = void 0;
exports.NO_UNUSED_NAMESPACE_TYPE = 'no-unused-namespace';
exports.NO_UNDEFINED_NAMESPACE_TYPE = 'no-undefined-namespace';
exports.NAME_CASE_ISSUE_PATH_VALUE = 'name-case-issue-path-value';
exports.MISSING_I18N_KEY = 'missing-i18n-key';
exports.VALUE_REQUIRED = 'value-required';
exports.INCOMPLETE_EXPRESSION_CC_FORWARD_SLASH = 'incomplete-expression-cc-forward-slash';
exports.INCOMPLETE_EXPRESSION_FORWARD_SLASH = 'incomplete-expression-forward-slash';
exports.IGNORE_TARGET_VALIDATION = 'ignore-target-validation';
exports.UNKNOWN_TERM = 'unknown-term';
exports.UN_SUPPORTED_VOCABULARY = 'un-supported-vocabulary';
exports.ATTRIBUTE_NOT_ALLOWED_HERE = 'attribute-not-allowed-here';
exports.MISSING_REQUIRED_ATTRIBUTE = 'missing-required-attribute';
exports.MISSING_REQUIRED_VALUE_FOR_ATTRIBUTE = 'missing-required-value-for-attribute';
exports.TERM_NOT_APPLICABLE = 'term-not-applicable';
exports.NOT_IN_APPLICABLE_TERMS_CONSTRAINT = 'not-in-applicable-terms-constraint';
exports.RECORD_COLLECTION_PATH_NOT_ALLOWED = 'record-collection-path-not-allowed';
exports.ODATA_FUNCTION_WRONG_RETURN_TYPE = 'odata-function-wrong-return-type';
exports.IGNORE_DUPLICATE = 'ignore-duplicate';
exports.INVALID_PATH_EXPRESSION = 'invlid-path-expression';
exports.INVALID_ENUM_MEMBER_TYPE = 'unknown-enum-member';
exports.INVALID_TYPE_TYPE = 'invalid-type';
exports.NO_VALIDATION_FOR_SUBNODES = 'no-validation-subnodes';
exports.INCOMPLETE_PATH_WITH_TYPE = 'incomplete-path-with-type';
exports.INCOMPLETE_PATH_WITH_COMPATIBLE_TYPES = 'incomplete-path-with-compatible-types';
exports.COMMON_CASE_ISSUE = 'common-case-issue';
exports.ODATA_PATH_SEPARATOR_RULE = 'no-odata-path-separator';
exports.INVALID_PRIMITIVE_TYPE = 'invalid-primitive-type';
