// @sap/ux-specification/dist/types/src/v2/controls/FilterBar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterPathType = exports.DefaultDateRangeValueType = exports.DateRangeType = void 0;
var DateRangeType;
(function (DateRangeType) {
    DateRangeType["DATERANGE"] = "DATERANGE";
    DateRangeType["DATE"] = "DATE";
    DateRangeType["FROM"] = "FROM";
    DateRangeType["TO"] = "TO";
    DateRangeType["DAYS"] = "DAYS";
    DateRangeType["LASTDAYS"] = "LASTDAYS";
    DateRangeType["LASTWEEKS"] = "LASTWEEKS";
    DateRangeType["WEEK"] = "WEEK";
    DateRangeType["LASTMONTHS"] = "LASTMONTHS";
    DateRangeType["MONTH"] = "MONTH";
    DateRangeType["QUARTER"] = "QUARTER";
    DateRangeType["LASTQUARTERS"] = "LASTQUARTERS";
    DateRangeType["LASTYEARS"] = "LASTYEARS";
    DateRangeType["LASTYEAR"] = "LASTYEAR";
    DateRangeType["YEAR"] = "YEAR";
    DateRangeType["NEXTDAYS"] = "NEXTDAYS";
    DateRangeType["NEXTWEEKS"] = "NEXTWEEKS";
    DateRangeType["NEXTMONTHS"] = "NEXTMONTHS";
    DateRangeType["NEXTQUARTERS"] = "NEXTQUARTERS";
    DateRangeType["NEXTYEARS"] = "NEXTYEARS";
    DateRangeType["NEXT"] = "NEXT";
    DateRangeType["SPECIFICMONTH"] = "SPECIFICMONTH";
    DateRangeType["YESTERDAY"] = "YESTERDAY";
    DateRangeType["YEARTODATE"] = "YEARTODATE";
    DateRangeType["TODAY"] = "TODAY";
    DateRangeType["TOMORROW"] = "TOMORROW";
    DateRangeType["THISWEEK"] = "THISWEEK";
    DateRangeType["LASTWEEK"] = "LASTWEEK";
    DateRangeType["LAST2WEEKS"] = "LAST2WEEKS";
    DateRangeType["LAST3WEEKS"] = "LAST3WEEKS";
    DateRangeType["LAST4WEEKS"] = "LAST4WEEKS";
    DateRangeType["LAST5WEEKS"] = "LAST5WEEKS";
    DateRangeType["NEXTWEEK"] = "NEXTWEEK";
    DateRangeType["NEXT2WEEKS"] = "NEXT2WEEKS";
    DateRangeType["NEXT3WEEKS"] = "NEXT3WEEKS";
    DateRangeType["NEXT4WEEKS"] = "NEXT4WEEKS";
    DateRangeType["NEXT5WEEKS"] = "NEXT5WEEKS";
    DateRangeType["THISMONTH"] = "THISMONTH";
    DateRangeType["LASTMONTH"] = "LASTMONTH";
    DateRangeType["NEXTMONTH"] = "NEXTMONTH";
    DateRangeType["THISQUARTER"] = "THISQUARTER";
    DateRangeType["LASTQUARTER"] = "LASTQUARTER";
    DateRangeType["NEXTQUARTER"] = "NEXTQUARTER";
    DateRangeType["QUARTER1"] = "QUARTER1";
    DateRangeType["QUARTER2"] = "QUARTER2";
    DateRangeType["QUARTER3"] = "QUARTER3";
    DateRangeType["QUARTER4"] = "QUARTER4";
    DateRangeType["TODAYFROMTO"] = "TODAYFROMTO";
})(DateRangeType || (exports.DateRangeType = DateRangeType = {}));
var DefaultDateRangeValueType;
(function (DefaultDateRangeValueType) {
    DefaultDateRangeValueType["YESTERDAY"] = "YESTERDAY";
    DefaultDateRangeValueType["TODAY"] = "TODAY";
    DefaultDateRangeValueType["THISWEEK"] = "THISWEEK";
    DefaultDateRangeValueType["LASTWEEK"] = "LASTWEEK";
    DefaultDateRangeValueType["THISMONTH"] = "THISMONTH";
    DefaultDateRangeValueType["TOMORROW"] = "TOMORROW";
    DefaultDateRangeValueType["LASTMONTH"] = "LASTMONTH";
    DefaultDateRangeValueType["THISQUARTER"] = "THISQUARTER";
    DefaultDateRangeValueType["LASTQUARTER"] = "LASTQUARTER";
    DefaultDateRangeValueType["THISYEAR"] = "THISYEAR";
    DefaultDateRangeValueType["LASTYEAR"] = "LASTYEAR";
    DefaultDateRangeValueType["LAST2WEEKS"] = "LAST2WEEKS";
    DefaultDateRangeValueType["LAST3WEEKS"] = "LAST3WEEKS";
    DefaultDateRangeValueType["LAST4WEEKS"] = "LAST4WEEKS";
    DefaultDateRangeValueType["LAST5WEEKS"] = "LAST5WEEKS";
    DefaultDateRangeValueType["YEARTODATE"] = "YEARTODATE";
    DefaultDateRangeValueType["QUARTER1"] = "QUARTER1";
    DefaultDateRangeValueType["QUARTER2"] = "QUARTER2";
    DefaultDateRangeValueType["QUARTER3"] = "QUARTER3";
    DefaultDateRangeValueType["QUARTER4"] = "QUARTER4";
    DefaultDateRangeValueType["DATETOYEAR"] = "DATETOYEAR";
})(DefaultDateRangeValueType || (exports.DefaultDateRangeValueType = DefaultDateRangeValueType = {}));
var FilterPathType;
(function (FilterPathType) {
    FilterPathType["key"] = "key";
    FilterPathType["catgory"] = "category";
})(FilterPathType || (exports.FilterPathType = FilterPathType = {}));
