// @sap/ux-specification/dist/types/src/v4/controls/Table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultPathType = exports.InitialLoadType = exports.TableTypeV4 = exports.SelectionMode = void 0;
var SelectionMode;
(function (SelectionMode) {
    SelectionMode["Multi"] = "Multi";
    SelectionMode["None"] = "None";
    SelectionMode["Single"] = "Single";
    SelectionMode["Auto"] = "Auto";
})(SelectionMode || (exports.SelectionMode = SelectionMode = {}));
var TableTypeV4;
(function (TableTypeV4) {
    TableTypeV4["ResponsiveTable"] = "ResponsiveTable";
    TableTypeV4["GridTable"] = "GridTable";
    TableTypeV4["AnalyticalTable"] = "AnalyticalTable";
    TableTypeV4["TreeTable"] = "TreeTable";
})(TableTypeV4 || (exports.TableTypeV4 = TableTypeV4 = {}));
var InitialLoadType;
(function (InitialLoadType) {
    InitialLoadType["Disabled"] = "Disabled";
    InitialLoadType["Enabled"] = "Enabled";
    InitialLoadType["Auto"] = "Auto";
})(InitialLoadType || (exports.InitialLoadType = InitialLoadType = {}));
var DefaultPathType;
(function (DefaultPathType) {
    DefaultPathType["Primary"] = "primary";
    DefaultPathType["Secondary"] = "secondary";
    DefaultPathType["Both"] = "both";
})(DefaultPathType || (exports.DefaultPathType = DefaultPathType = {}));
