// @sap/ux-specification/dist/types/src/apiTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Features = exports.CustomExtensionType = exports.TemplateType = exports.LogSeverityLabel = exports.LogSeverity = void 0;
var LogSeverity;
(function (LogSeverity) {
    LogSeverity["Error"] = "error";
    LogSeverity["Warning"] = "warning";
    LogSeverity["Info"] = "info";
})(LogSeverity || (exports.LogSeverity = LogSeverity = {}));
exports.LogSeverityLabel = {
    ["error" /* LogSeverity.Error */]: 'Error',
    ["warning" /* LogSeverity.Warning */]: 'Warning',
    ["info" /* LogSeverity.Info */]: 'Information'
};
var TemplateType;
(function (TemplateType) {
    TemplateType["ListReportObjectPageV2"] = "ListReportObjectPageV2";
    TemplateType["ListReportObjectPageV4"] = "ListReportObjectPageV4";
    TemplateType["OverviewPageV2"] = "OverviewPageV2";
    TemplateType["AnalyticalListPageV2"] = "AnalyticalListPageV2";
    TemplateType["AnalyticalListPageV4"] = "AnalyticalListPageV4";
    TemplateType["FreestylePageV4"] = "FreestylePageV4";
})(TemplateType || (exports.TemplateType = TemplateType = {}));
var CustomExtensionType;
(function (CustomExtensionType) {
    CustomExtensionType["ControllerExtension"] = "ControllerExtension";
    CustomExtensionType["CustomAction"] = "CustomAction";
    CustomExtensionType["CustomColumn"] = "CustomColumn";
    CustomExtensionType["CustomFilterField"] = "CustomFilterField";
    CustomExtensionType["CustomHeaderSection"] = "CustomHeaderSection";
    CustomExtensionType["CustomPage"] = "CustomPage";
    CustomExtensionType["CustomSection"] = "CustomSection";
    CustomExtensionType["CustomSubSection"] = "CustomSubSection";
    CustomExtensionType["CustomView"] = "CustomView";
    CustomExtensionType["ListReport"] = "ListReport";
    CustomExtensionType["ObjectPage"] = "ObjectPage";
})(CustomExtensionType || (exports.CustomExtensionType = CustomExtensionType = {}));
/**
 * API Feature toggles
 */
var Features;
(function (Features) {
    Features["BuildingBlocks"] = "BuildingBlocks";
})(Features || (exports.Features = Features = {}));
