// @sap-ux/fiori-annotation-api/dist/xml/changes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.insertElement = exports.MOVE_COLLECTION_VALUE = exports.DELETE_ATTRIBUTE = exports.UPDATE_ATTRIBUTE_VALUE = exports.UPDATE_ATTRIBUTE_NAME = exports.INSERT_ATTRIBUTE = exports.DELETE_ELEMENT = exports.UPDATE_ELEMENT_NAME = exports.REPLACE_ELEMENT_CONTENT = exports.REPLACE_ELEMENT = exports.INSERT_ELEMENT = void 0;
exports.INSERT_ELEMENT = 'insert-element';
exports.REPLACE_ELEMENT = 'replace-element';
exports.REPLACE_ELEMENT_CONTENT = 'replace-element-content';
exports.UPDATE_ELEMENT_NAME = 'update-element-name';
exports.DELETE_ELEMENT = 'delete-element';
exports.INSERT_ATTRIBUTE = 'insert-attribute';
exports.UPDATE_ATTRIBUTE_NAME = 'update-attribute-name';
exports.UPDATE_ATTRIBUTE_VALUE = 'update-attribute-value';
exports.DELETE_ATTRIBUTE = 'delete-attribute';
exports.MOVE_COLLECTION_VALUE = 'move-collection-value';
/**
 *  Creates {@link InsertElement} change.
 *
 * @param pointer - Pointer to the element.
 * @param element - Value to be inserted.
 * @param index - Position before which element should the new value be inserted.
 * @returns InsertElement change object.
 */
function insertElement(pointer, element, index) {
    const change = {
        type: 'insert-element',
        pointer,
        element
    };
    if (index !== undefined) {
        change.index = index;
    }
    return change;
}
exports.insertElement = insertElement;
