// @sap/ux-specification/dist/types/src/common/webapp/manifest/Manifest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColorPaletteDefinitionType = exports.GENERICAPPSETTINGS = exports.ManifestSection = exports.DataSourceType = void 0;
var DataSourceType;
(function (DataSourceType) {
    DataSourceType["OData"] = "OData";
    DataSourceType["ODataAnnotation"] = "ODataAnnotation";
})(DataSourceType || (exports.DataSourceType = DataSourceType = {}));
var ManifestSection;
(function (ManifestSection) {
    ManifestSection["ui"] = "sap.ui";
    ManifestSection["app"] = "sap.app";
    ManifestSection["generic"] = "sap.ui.generic.app";
    ManifestSection["ovp"] = "sap.ovp";
    ManifestSection["ui5"] = "sap.ui5";
})(ManifestSection || (exports.ManifestSection = ManifestSection = {}));
exports.GENERICAPPSETTINGS = `${ManifestSection.generic}/settings`;
var ColorPaletteDefinitionType;
(function (ColorPaletteDefinitionType) {
    ColorPaletteDefinitionType["SemanticColoring"] = "SemanticColoring";
    ColorPaletteDefinitionType["StableColoring"] = "StableColoring";
    ColorPaletteDefinitionType["MapDimensionValues"] = "MapDimensionValues";
})(ColorPaletteDefinitionType || (exports.ColorPaletteDefinitionType = ColorPaletteDefinitionType = {}));
