// @sap-ux/fiori-annotation-api/dist/types/internal-change.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.REPLACE_TEXT = exports.MOVE_ELEMENT = exports.DELETE_ELEMENT = exports.DELETE_ATTRIBUTE = exports.REPLACE_ELEMENT = exports.REPLACE_ATTRIBUTE = exports.REPLACE_ELEMENT_CONTENT = exports.UPDATE_ELEMENT_NAME = exports.UPDATE_ATTRIBUTE_VALUE = exports.INSERT_ATTRIBUTE = exports.INSERT_ELEMENT = exports.INSERT_TARGET = exports.DELETE_REFERENCE = void 0;
// change objects describing edits in generic format
exports.DELETE_REFERENCE = 'delete-reference';
exports.INSERT_TARGET = 'insert-target';
exports.INSERT_ELEMENT = 'insert-element';
exports.INSERT_ATTRIBUTE = 'insert-attribute';
exports.UPDATE_ATTRIBUTE_VALUE = 'update-attribute-value';
exports.UPDATE_ELEMENT_NAME = 'update-element-name';
exports.REPLACE_ELEMENT_CONTENT = 'replace-element-content';
exports.REPLACE_ATTRIBUTE = 'replace-attribute';
exports.REPLACE_ELEMENT = 'replace-element';
exports.DELETE_ATTRIBUTE = 'delete-attribute';
exports.DELETE_ELEMENT = 'delete-element';
exports.MOVE_ELEMENT = 'move-element';
exports.REPLACE_TEXT = 'replace-text';
